raspberry/main.py

```diff
@@ -0,0 +1,371 @@
+        2. Updates shared variable values using radio data.
+        3. Calls cache_values function.
+        """
+        # if altitude hold is enabled and drone is not close to the ground update the altitude_ref_mts
+        if use_altitude_hold[0] and not is_drone_flying_close_to_ground[0]:
+            altitude_ref_mts[0] = current_altitude_snap_shot_mts[0] + (
+                var_e_RC_mid_percentage[0] - rc.trimmer_VRE_percentage())*altitude_shifter_range_mts[0]
+            # the altitude_ref_mts should not be less than minimum flight altitude
+            altitude_ref_mts[0] = max(
+                min_altitude_to_activate_AltiHold_mts[0], altitude_ref_mts[0])
+
+        # read, encode, and send the radio data to ESP.
+        # if altitude hold is on the thorttle value from the RC will be overwritten
+        # by the throttle reference from the LQR.
+        # NOTE: There is this weird conversion for LQR throttle reference below,
+        # this is because the PI needs to send throttle reference in the range [300, 1400],
+        # which is the actual range of the RC throttle stick.
+        shit = int((throttle_ref_from_LQR[0][0][0] - 1000) * 1400/900 +
+                   300) if use_altitude_hold[0] and not is_drone_flying_close_to_ground[0] else -1
+        channel_data[0] = rc.get_radio_data_parse_and_send_to_ESP(return_channel_date=True,
+                                                                  force_send_fake_data=False,
+                                                                  fake_data="S,0,0,0,0,0,0,0,0,0",
+                                                                  over_write_throttle_ref_to=shit)
+        # update shared variables using RC data
+        # is data logging killed and data saving requested?
+        is_data_log_kill[0] = rc.switch_A()
+        # is altitude hold enabled?
+        use_altitude_hold[0] = rc.switch_C()
+        LQR_Q11_gain[0] = rc.trimmer_VRA_percentage(
+        )*LQR_Q11_GAIN_MAX  # Kappa11 gain from RC
+        LQR_Q22_gain[0] = rc.trimmer_VRB_percentage(
+        )*LQR_Q22_GAIN_MAX  # kappa22 gain from RC
+        LQR_R_gain[0] = rc.trimmer_VRC_percentage() * LQR_R_GAIN_MAX + \
+            1  # R data for LQR from RC
+        # normalised throttle reference from RC, max is 1900
+        Tref_t[0] = (rc.throttle_reference_percentage() - 1000)/900
+        cache_values()  # call to cache values
+
+    def process_ESP_data():
+        """Process ESP data, this function does two jobs:
+        1. Receives flight data as a string of space seperated values formatted as "FD: q1 q2 q3 ax ay az".
+        2. Checks the received flight data for corruption. 
+           If string is not None and starts with "FD:", and if there are 7 space seperated values
+           then convert the 7 values to floats and update shared variables.
+        NOTE: might want to include a check for value data corruption before converting to floats, 
```

Comment:
Why don't you make an issue about this? Otherwise, we'll forget.

Reply:
Done! Issue #108 opened to address this.